### PR TITLE
Switch is_contiguous to use guardless implementation

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #92234
* #94410
* #94408
* #94407
* #94404
* #94400
* #94399
* #94201
* #94307

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
That is pretty tricky logic but change sounds good.
Do we want to test to make sure we will check regression if guards are added in the future?

**Comment by ezyang** [comment]
The test will probably be that you can construct tensors with unbacked symint

**Review by albanD — APPROVED** [review]
Merging as-is sounds good and this will be tested when the unbaked symint is fully online.

**Comment by albanD** [comment]
Some tests need updating though.

File: c10/core/TensorImpl.cpp [hunk 1]
```diff
@@ -292,33 +300,44 @@ bool _compute_channels_last_contiguous_2d(
...
+        const auto& size_d = sizes[d];
+        r = r & (size_d.sym_eq(1) | strides[d].sym_eq(expected));
+        expected *= size_d;
+      }
+      return r;
+    }
+    // NOLINTNEXTLINE(bugprone-branch-clone)
+    case 3:
+      // TODO dim == 3 case will be enabled once it is fully tested
+      return false;
```
Comment (bdhirsh) [inline c10/core/TensorImpl.cpp:324]: hmm what does this mean? (Why do we always want to report this case as non-contiguous until it's fully tested?)

File: c10/core/TensorImpl.cpp [hunk 2]
```diff
@@ -429,21 +513,50 @@ bool _compute_non_overlapping_and_dense(
...
+  if (!base) {
+    for (const auto& s : strides) {
+      if (s.is_symbolic()) {
+        base = s.toSymNodeImpl();
+        break;
+      }
+    }
+  }
+  if (!base) {
+    // Do the normal computation
```
Comment (bdhirsh) [inline c10/core/TensorImpl.cpp:540]: for my understanding - when you can you end up in a situation where you call the `identity<SymBool>` variant of this method, but not have any actual SymInts stored on the TensorImpl to dispatch on?

File: c10/core/TensorImpl.cpp (hunk 1 above)
Comment (ezyang) [reply c10/core/TensorImpl.cpp:324]: This is from the old code. What I believe happened was the original memory format workstream was going to change the definition of contiguity to also handle the dim 3 case, but people never got around to doing it.

File: c10/core/TensorImpl.cpp (hunk 2 above)
Comment (ezyang) [reply c10/core/TensorImpl.cpp:540]: storage offset is symbolic, nothing else is symbolic.

File: c10/core/TensorImpl.cpp [hunk 3]
```diff
@@ -430,74 +514,89 @@ bool _compute_non_overlapping_and_dense(
...
+    return compute_non_overlapping_and_dense(identity<bool>());
```
Comment (ezyang) [inline c10/core/TensorImpl.cpp:542]: This is actually wrong 🤔  All the metadata is on `extra_meta_` but this is pulled from the wrong place

**Comment by ezyang** [comment]
this is dead, redone in https://github.com/pytorch/pytorch/pull/94473/

**Pull request closed without merging**